crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,359 @@ impl<'a> FileNoqaDirectives<'a> {
+            source,
+            TextRange::new(offset + noqa_literal_end, comment_range.end()),
+        );
+        return Ok(Some(lexer.lex(offset + comment_start)?));
     }
 
-    /// Lex optional leading whitespace.
-    #[inline]
-    fn lex_whitespace(line: &str) -> &str {
-        line.trim_start()
+    Ok(None)
+}
+
+/// Helper to check if "noqa" (case-insensitive) appears at the given position
+#[inline]
+fn is_noqa_at_position(text: &str, pos: usize) -> bool {
+    matches!(
+        text[pos..].as_bytes(),
+        [b'n' | b'N', b'o' | b'O', b'q' | b'Q', b'a' | b'A', ..]
+    )
+}
+
+pub(crate) fn lex_codes(text: &str) -> Result<Vec<Code<'_>>, LexicalError> {
+    let mut lexer = NoqaLexer::starts_at(TextSize::new(0), text);
+    lexer.lex_codes()?;
+    Ok(lexer.codes)
+}
+
+/// Lexer for `noqa` comment lines.
+///
+/// Assumed to be initialized with range or offset that begins
+/// at or after `noqa` literal, e.g. the ":" in `# noqa: F401`.
+#[derive(Debug)]
+struct NoqaLexer<'a> {
+    /// A slice of source text starting at the end of the `noqa` literal
+    /// e.g. `: F401` in `# noqa: F401`
+    line: &'a str,
+    /// Contains convenience methods for lexing
+    cursor: Cursor<'a>,
+    /// Byte offset of end of `noqa` literal in source text
+    offset: TextSize,
+    /// Non-fatal warnings collected during lexing
+    warnings: Vec<LexicalWarning>,
+    /// Tracks whether we are lexing in a context with a missing delimiter
+    /// e.g. at `C` in `F401C402`.
+    missing_delimiter: bool,
+    /// Rule codes collected during lexing
+    codes: Vec<Code<'a>>,
+}
+
+impl<'a> NoqaLexer<'a> {
+    fn new(source: &'a str, range: TextRange) -> Self {
+        Self {
+            line: &source[range],
+            offset: range.start(),
+            cursor: Cursor::new(&source[range]),
+            warnings: Vec::new(),
+            missing_delimiter: false,
+            codes: Vec::new(),
+        }
     }
 
-    /// Lex a specific character, or return `None` if the character is not the first character in
-    /// the line.
-    #[inline]
-    fn lex_char(line: &str, c: char) -> Option<&str> {
-        let mut chars = line.chars();
-        if chars.next() == Some(c) {
-            Some(chars.as_str())
-        } else {
-            None
+    /// Initialize [`NoqaLexer`] at offset.
+    ///
+    /// The offset should be at or after the `noqa` literal.
+    fn starts_at(offset: TextSize, source: &'a str) -> Self {
+        let range = TextRange::new(offset, source.text_len());
+        Self::new(source, range)
+    }
+
+    /// Collect codes in `noqa` comment.
+    ///
+    /// The `comment_start` is the offset in the source of the
+    /// beginning of the `noqa` comment, which is earlier
+    /// than the offset that marks the beginning of the codes.
+    fn lex(mut self, comment_start: TextSize) -> Result<NoqaLexerOutput<'a>, LexicalError> {
+        if self.cursor.first().is_ascii_whitespace()
```

Comment:
It's not clear to me why we'd return `All` if the lexer is at a whitespace. What if we have `# noqa:        RUF100`? Would that return `All`?

Should this use `is_whitespace` similar to how you use `trim_end` in other places?